commands/download.go

```diff
@@ -1,15 +1,119 @@
 package commands
 
 import (
+	"context"
+	"errors"
 	"fmt"
-	"github.com/jfrog/jfrog-cli-core/plugins/components"
-	"github.com/jfrog/jfrog-cli-core/utils/config"
+	"github.com/jfrog/jfrog-client-go/utils/io/fileutils"
 	"github.com/jfrog/jfrog-client-go/utils/log"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
 )
 
-//nolint: unparam, lll
-func downloadSupportBundle(ctx *components.Context, rtDetails *config.ArtifactoryDetails, conf *supportBundleCommandConfiguration, response creationResponse) (string, error) {
-	tmpFile := "/tmp/foo.tmp"
-	log.Debug(fmt.Sprintf("Download Support Bundle %s from %s to %s", response.ID, rtDetails.Url, tmpFile))
-	return tmpFile, nil
+type downloadSupportBundleHttpClient interface {
+	GetURL() string
+	DownloadSupportBundle(bundleId string) (*http.Response, error)
+	GetSupportBundleStatus(bundleId string) (int, []byte, error)
+}
+
+func downloadSupportBundle(client downloadSupportBundleHttpClient, timeout time.Duration, retryInterval time.Duration,
+	bundleId string) (string, error) {
+
+	dirPath, err := fileutils.CreateTempDir()
+	if err != nil {
+		return "", err
+	}
+	tmpFilePath := filepath.Join(dirPath, fmt.Sprintf("%s.zip", bundleId))
+	tmpZipFile, err := os.Create(tmpFilePath)
+	defer handleClose(tmpZipFile)
+	if err != nil {
+		return "", err
+	}
+	log.Debug(fmt.Sprintf("Download Support Bundle %s from %s to %s", bundleId, client.GetURL(), tmpFilePath))
+
+	err = waitUntilSupportBundleIsReady(client, retryInterval, timeout, bundleId)
+	if err != nil {
+		return "", err
+	}
+
+	err = downloadSupportBundleAndWriteToFile(client, tmpZipFile, bundleId)
+	if err != nil {
+		return "", err
+	}
+
+	return tmpFilePath, nil
+}
+
+func downloadSupportBundleAndWriteToFile(client downloadSupportBundleHttpClient, tmpZipFile *os.File, bundleId string) error {
+	resp, err := client.DownloadSupportBundle(bundleId)
+	if resp != nil {
+		defer handleClose(resp.Body)
+	}
+	if err != nil {
+		return err
+	}
+	log.Debug(fmt.Sprintf("Got %d", resp.StatusCode))
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("http request failed with: %d %s", resp.StatusCode, http.StatusText(resp.StatusCode))
+	}
+
+	_, err = io.Copy(tmpZipFile, resp.Body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func waitUntilSupportBundleIsReady(client downloadSupportBundleHttpClient, retryInterval time.Duration,
+	timeout time.Duration, bundleId string) error {
+
+	ctx, cancelCtx := context.WithTimeout(context.Background(), timeout)
```

Comment:
We should get `context.Background()` in the command and pass it to each phase that needs it. Otherwise, we'll end up with multiple scattered `context.Background()` calls.

Reply:
I agree, I just put it here because it wasn't used elsewhere. I'll move it.

Reply:
Let's do that in another PR, yours is already big enough :)

Reply:
It wasn't a very big change.